app/scripts/migrations/099.ts

```diff
@@ -0,0 +1,116 @@
+import {
+  EthAccountType,
+  InternalAccount,
+  EthMethod,
+} from '@metamask/keyring-api';
+import { sha256FromString } from 'ethereumjs-util';
+import { v4 as uuid } from 'uuid';
+import { cloneDeep } from 'lodash';
+
+type VersionedData = {
+  meta: { version: number };
+  data: Record<string, unknown>;
+};
+
+export const version = 99;
+
+/**
+ * This migration does the following:
+ *
+ * - Creates a default state for AccountsController.
+ * - Moves identites and selectedAddress from the PreferencesController to the AccountsController state.
+ * - Removes identites and selectedAddress from the PreferencesController
+ *
+ * @param originalVersionedData - Versioned MetaMask extension state, exactly what we persist to dist.
+ * @param originalVersionedData.meta - State metadata.
+ * @param originalVersionedData.meta.version - The current state version.
+ * @param originalVersionedData.data - The persisted MetaMask state, keyed by controller.
+ * @returns Updated versioned MetaMask extension state.
+ */
+export async function migrate(
+  originalVersionedData: VersionedData,
+): Promise<VersionedData> {
+  const versionedData = cloneDeep(originalVersionedData);
+  versionedData.meta.version = version;
+  migrateData(versionedData.data);
+  return versionedData;
+}
+
+function migrateData(state: Record<string, unknown>): void {
+  createDefaultAccountsController(state);
+  moveIdentitiesToAccountsController(state);
+  moveSelectedAddressToAccountsController(state);
+  removeIdentitiesAndSelectedAddressFromPreferencesController(state);
+}
+
+function createDefaultAccountsController(state: Record<string, any>) {
+  state.AccountsController = {
+    internalAccounts: {
+      accounts: {},
+      selectedAccount: '',
+    },
+  };
+}
+
+function moveIdentitiesToAccountsController(state: Record<string, any>) {
+  const identities: {
+    [key: string]: {
+      name: string;
+      address: string;
+      lastSelected: number;
+    };
+  } = state.PreferencesController?.identities || {};
+
+  if (Object.keys(identities).length === 0) {
+    return;
+  }
+
+  const accounts: Record<string, InternalAccount> = {};
+
+  Object.values(identities).forEach((identity) => {
+    const expectedId = uuid({
+      random: sha256FromString(identity.address).slice(0, 16),
+    });
+
+    accounts[expectedId] = {
+      address: identity.address,
+      id: expectedId,
+      options: {},
+      metadata: {
+        name: identity.name,
+        lastSelected: identity.lastSelected ?? null,
+        keyring: {
+          type: 'HD Key Tree',
```

Comment:
Add a comment about the keyring type being updated on the first load after the migration.

Reply:
Added on commit 6c3dccac787d4840192d4bf9fa596075a889f36f